cg/smart_contracts/ClearGDPR.sol

```diff
@@ -1,135 +1,152 @@
-    }
-    for(i = 0; i < _processorsWhiteListedBySubject.length; i++){
-      subjects[_subjectIdHash].statePerProcessor[_processorsWhiteListedBySubject[i]] = State.shareable;  
-    }
-    subjects[_subjectIdHash].isErased = false;
-    emit Controller_ConsentGivenTo(_subjectIdHash, _processorsWhiteListedBySubject);
-    return true;
-  }
+    function recordConsentGivenTo(bytes32 _subjectId, address[] _processorsConsented) public onlyController notErased(_subjectId) returns(bool){
+        require(areAllValidProcessors(_processorsConsented));
+        for(uint256 i = 0; i < processors.length; i++){
+            subjects[_subjectId].statePerProcessor[processors[i]] = State.notShareable;  
+        }
+        for(i = 0; i < _processorsConsented.length; i++){
+            subjects[_subjectId].statePerProcessor[_processorsConsented[i]] = State.shareable;  
+        }
+        subjects[_subjectId].statePerProcessor[controller] = State.shareable;  
+        subjects[_subjectId].isErased = false;
+        emit Controller_ConsentGivenTo(_subjectId, _processorsConsented);
+        return true;
+    }
   
-  function recordAccessByController(bytes32 _subjectIdHash) public onlyController notErased(_subjectIdHash) returns (State[]){
-    State[] memory states = new State[](processors.length);
-    for(uint256 i = 0; i < processors.length; i++){
-      states[i] = subjects[_subjectIdHash].statePerProcessor[processors[i]];
-    }
-    emit Controller_SubjectDataAccessed(_subjectIdHash);
-    return states;
-  }
+    function recordAccessByController(bytes32 _subjectId) public onlyController notErased(_subjectId) returns(State[]){
+        State[] memory states = new State[](processors.length);
+        for(uint256 i = 0; i < processors.length; i++){
+            states[i] = subjects[_subjectId].statePerProcessor[processors[i]];
+        }
+        emit Controller_SubjectDataAccessed(_subjectId);
+        return states;
+    }
   
-  function recordRectificationByController(bytes32 _subjectIdHash) public onlyController notErased(_subjectIdHash) returns (bool){
-    subjects[_subjectIdHash].rectificationCount++;
-    emit Controller_SubjectDataRectified(_subjectIdHash, subjects[_subjectIdHash].rectificationCount);
-    return true;
-  }
+    function recordRectificationByController(bytes32 _subjectId) public onlyController notErased(_subjectId) returns(bool){
+        subjects[_subjectId].rectificationCount++;
+        emit Controller_SubjectDataRectified(_subjectId, subjects[_subjectId].rectificationCount);
+        return true;
+    }
 
-  function recordErasureByController(bytes32 _subjectIdHash) public onlyController notErased(_subjectIdHash) returns (bool){
-    subjects[_subjectIdHash].isErased = true;
-    emit Controller_SubjectDataErased(_subjectIdHash);
-    return true;
-  }
+    function recordErasureByController(bytes32 _subjectId) public onlyController notErased(_subjectId) returns(bool){
+        subjects[_subjectId].isErased = true;
```

Comment:
How about this line above?

```
subjects[_subjectId].statePerProcessor[controller] = State.shareable;
```

Don't we also need to set erased state in the `statePerProcessor` array?

Reply:
No because so far that's a property of the subject's data. We still had not considered the case that a processor deletes his copy of the data, without a request to be forgotten being issued by the subject. We should create an issue to handle this case in the future

Reply:
In the right to restrict I will have a good setup to handle this